ee/modules/021-cni-cilium/hooks/ee/egressgateway_discovery.go

```diff
@@ -0,0 +1,345 @@
+/*
+Copyright 2024 Flant JSC
+Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+*/
+
+package ee
+
+import (
+	"fmt"
+	eeCrd "github.com/deckhouse/deckhouse/ee/modules/021-cni-cilium/hooks/ee/lib/crd"
+	"github.com/deckhouse/deckhouse/go_lib/dependency"
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"strings"
+)
+
+const (
+	cniNamespace       = "d8-cni-cilium"
+	memberLabelKey     = "egress-gateway.network.deckhouse.io/member"
+	masterLabelPreffix = "egress-gateway.network.deckhouse.io/master/"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+	Queue:        "/modules/cni-cilium/egress-discovery",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       "egressgatewayfailovergroups",
+			ApiVersion: "network.deckhouse.io/v1alpha1",
+			Kind:       "EgressGatewayFailoverGroup",
+			FilterFunc: applyEgressGatewayFilter,
+		},
+		{
+			Name:       "nodes",
+			ApiVersion: "v1",
+			Kind:       "Node",
+			FilterFunc: applyNodeFilter,
+		},
+		{
+			Name:       "pods",
+			ApiVersion: "v1",
+			Kind:       "Pod",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{cniNamespace},
+				},
+			},
+			NameSelector: &types.NameSelector{MatchNames: []string{
+				"agent",
+			}},
+			FilterFunc: applyPodFilter,
+		},
+	},
+	Schedule: []go_hook.ScheduleConfig{
+		{Name: "cron", Crontab: "* * * * *"},
+	},
+}, dependency.WithExternalDependencies(handleEgressGateways))
+
+func applyEgressGatewayFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var eg eeCrd.EgressGatewayFailoverGroup
+
+	err := sdk.FromUnstructured(obj, &eg)
+	if err != nil {
+		return nil, err
+	}
+
+	//return EgressGatewayGroupInfo{
+	//	Name:               eg.Name,
+	//	NodeSelector:       eg.Spec.NodeSelector,
+	//	AllNodes:           make([]string, 0, 4),
+	//	ReadyNodes:         make([]string, 0, 4),
+	//	CurrentMasterNodes: make([]string, 0, 4),
+	//}, nil
+	return eg, nil
```

Comment:
It is important to return Info.
Separate the AllNodes and other arrays to other struct like EgressGatewayFailoverGroupState.